### ISSUE TITLE
Release v0.10

### ISSUE THREAD

**Issue by BigLep** [ipfs/go-ipfs#8176]
# go-ipfs 0.10.0 Release

We're happy to announce go-ipfs 0.10.0. This release brings some big changes to the IPLD internals of go-ipfs that make working with non-UnixFS DAGs easier than ever. There are also a variety of new commands and configuration options available.

As usual, this release includes important fixes, some of which may be critical for security. Unless the fix addresses a bug being exploited in the wild, the fix will _not_ be called out in the release notes. Please make sure to update ASAP. See our [release process](https://github.com/ipfs/go-ipfs/tree/master/docs/releases.md#security-fix-policy) for details.

# 🛠 TLDR: BREAKING CHANGES

- `ipfs dag get`
  - default output changed to [`dag-json`](https://ipld.io/specs/codecs/dag-json/spec/)
  - dag-pb (e.g. unixfs) field names changed - impacts userland code that works with `dag-pb` objects returned by `dag get`
  - no longer emits an additional new-line character at the end of the data output
- `ipfs dag put`
  - defaults changed to reduce ambiguity and surprises: input is now assumed to be [`dag-json`](https://ipld.io/specs/codecs/dag-json/spec/), and data is serialized to [`dag-cbor`](https://ipld.io/specs/codecs/dag-cbor/spec/) at rest.
  - `--format` and `--input-enc` were removed and replaced with `--store-codec` and `--input-codec`
  - codec names now match the ones defined in the [multicodec table](https://github.com/multiformats/multicodec/blob/master/table.csv) 
  - dag-pb (e.g. unixfs) field names changed - impacts userland code that works with `dag-pb` objects stored via `dag put`

Keep reading to learn more details.

# 🔦 Highlights

## 🌲 IPLD Levels Up

The handling of data serialization as well as many aspects of DAG traversal and pathing have been migrated from older libraries, including [go-merkledag](https://github.com/ipfs/go-merkledag) and [go-ipld-format](https://github.com/ipfs/go-ipld-format) to the new **[go-ipld-prime](https://github.com/ipld/go-ipld-prime)** library and its components. This allows us to use many of the newer tools afforded by go-ipld-prime, stricter and more uniform codec implementations, support for additional (pluggable) codecs, and some minor performance improvements.

This is significant refactor of a core component that touches many parts of IPFS, and does come with some **breaking changes**:

* **IPLD plugins**:
  * The `PluginIPLD` interface has been changed to utilize go-ipld-prime. There is a demonstration of the change in the [bundled git plugin](./plugin/plugins/git/).
* **The semantics of `dag put` and `dag get` change**:
  * `dag get` now takes the `output-codec` option which accepts a [multicodec](https://docs.ipfs.io/concepts/glossary/#multicodec) name used to encode the output. By default this is `dag-json`, which is  a strict and deterministic subset of JSON created by the IPLD team. Users may notice differences from the previously plain Go JSON output, particularly where bytes are concerned which are now encoded using a form similar to CIDs: `{"/":{"bytes":"unpadded-base64-bytes"}}` rather than the previously Go-specific plain padded base64 string. See the [dag-json specification](https://ipld.io/specs/codecs/dag-json/spec/) for an explanation of these forms.
  * `dag get` no longer prints an additional new-line character at the end of the encoded block output. This means that the output as presented by `dag get` are the exact bytes of the requested node. A round-trip of such bytes back in through `dag put` using the same codec should result in the same CID.
  * `dag put` uses the `input-codec` option to specify the multicodec name of the format data is being provided in, and the `store-codec` option to specify the multicodec name of the format the data should be stored in at rest. These formerly defaulted to `json` and `cbor` respectively. They now default to `dag-json` and `dag-cbor` respectively but may be changed to any supported codec (bundled or loaded via plugin) by its [multicodec name](https://github.com/multiformats/multicodec/blob/master/table.csv).
  * The `json` and `cbor` multicodec names (as used by `input-enc` and `format` options) are now no longer aliases for `dag-json` and `dag-cbor` respectively. Instead, they now refer to their proper [multicodec](https://github.com/multiformats/multicodec/blob/master/table.csv) types. `cbor` refers to a plain CBOR format, which will not encode CIDs and does not have strict deterministic encoding rules. `json` is a plain JSON format, which also won't encode CIDs and will encode bytes in the Go-specific padded base64 string format rather than the dag-json method of byte encoding. See https://ipld.io/specs/codecs/ for more information on IPLD codecs.
  * `protobuf` is no longer used as the codec name for `dag-pb`
  * The codec name `raw` is used to mean Bytes in the [IPLD Data Model](https://github.com/ipld/specs/blob/master/data-model-layer/data-model.md#bytes-kind)
* **UnixFS refactor**. The **dag-pb codec**, which is used to encode UnixFS data for IPFS, is now represented through the `dag` API in a form that mirrors the protobuf schema used to define the binary format. This unifies the implementations and specification of dag-pb across the IPLD and IPFS stacks. Previously, additional layers of code for file and directory handling within IPFS between protobuf serialization and UnixFS obscured the protobuf representation. Much of this code has now been replaced and there are fewer layers of transformation. This means that interacting with dag-pb data via the `dag` API will use different forms:
  * Previously, using `dag get` on a dag-pb block would present the block serialized as JSON as `{"data":"padded-base64-bytes","links":[{"Name":"foo","Size":100,"Cid":{"/":"Qm..."}},...]}`.
  * Now,  the dag-pb data with dag-json codec for output will be serialized using the data model from the [dag-pb specification](https://ipld.io/specs/codecs/dag-pb/spec/): `{"Data":{"/":{"bytes":"unpadded-base64-bytes"}},"Links":[{"Name":"foo","Tsize":100,"Hash":{"/":"Qm..."}},...]}`. Aside from the change in byte formatting, most field names have changed: `data` → `Data`, `links` → `Links`, `Size` → `Tsize`, `Cid` → `Hash`. Note that this output can be changed now using the `output-codec` option to specify an alternative codec.
  * Similarly, using `dag put` and a `store-codec` option of `dag-pb` now requires that the input conform to this dag-pb specified form. Previously, input using `{"data":"...","links":[...]}` was accepted, now it must be `{"Data":"...","Links":[...]}`.
  * Previously it was not possible to use paths to navigate to any of these properties of a dag-pb node, the only possible paths were named links, e.g. `dag get QmFoo/NamedLink` where `NamedLink` was one of the links whose name was `NamedLink`. This functionality remains the same, but by prefixing the path with `/ipld/` we enter data model pathing semantics and can `dag get /ipld/QmFoo/Links/0/Hash` to navigate to links or `/ipld/QmFoo/Data` to simply retrieve the data section of the node, for example.
  * ℹ See the [dag-pb specification](https://ipld.io/specs/codecs/dag-pb/) for details on the codec and its data model representation.
  * ℹ See this [detailed write-up](https://github.com/ipld/ipld/blob/master/design/tricky-choices/dag-pb-forms-impl-and-use.md) for further background on these changes.

## Ⓜ Multibase Command

go-ipfs now provides utility commands for working with [multibase](https://docs.ipfs.io/concepts/glossary/#multibase):

```console
$ echo -n hello | ipfs multibase encode -b base16 > file-mbase16
$ cat file-mbase16
f68656c6c6f

$ ipfs multibase decode file-mbase16
hello

$ cat file-mbase16 | ipfs multibase decode
hello

$ ipfs multibase transcode -b base2 file-mbase16
00110100001100101011011000110110001101111
```

See `ipfs multibase --help` for more examples.

## 🔨 Bitswap now supports greater configurability

This release adds an [`Internal` section](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#internal) to the configuration file that is designed to help advanced users optimize their setups without needing a custom binary. The `Internal` section is not guaranteed to be the same from release to release and may not be covered by migrations. If you use the `Internal` section you should be making sure to check the config documentation between releases for any changes.

## 🐚 Programmatic shell completions command

`ipfs commands completion bash` will generate a bash completion script for go-ipfs commands

## 📜 Profile collection command

Performance profiles can now be collected using `ipfs diag profile`. If you need to do some debugging or have an issue to submit the collected profiles are very useful to have around.

## 🍎 Mac OS notarized binaries

The go-ipfs and related migration binaries (for both Intel and Apple Sillicon) are now signed and notarized to make Mac OS installation easier.

## 👨‍👩‍👦 Improved MDNS

There is a completed implementation of the revised libp2p MDNS spec. This should result in better MDNS discovery and better local/offline operation as a result.

## 🚗 CAR import statistics

`dag import` command now  supports `--stats` option which will include the number of imported blocks and their total size in the output.

## 🕸 Peering command

This release adds `swarm peering`  command for easy management of  the peering subsystem. Peer in the peering subsystem is maintained to be connected at all times, and gets reconnected on disconnect with a back-off.

See `ipfs swarm peering --help` for more details.

# ✅ Release Checklist

For each RC published in each stage:

- version string in `version.go` has been updated (in the `release-vX.Y.Z` branch).
- tag commit with `vX.Y.Z-rcN`
- upload to dist.ipfs.io
  1. Build: https://github.com/ipfs/distributions#usage.
  2. Pin the resulting release.
  3. Make a PR against ipfs/distributions with the updated versions, including the new hash in the PR comment.
  4. Ask the infra team to update the DNSLink record for dist.ipfs.io to point to the new distribution.
- cut a pre-release on [github](https://github.com/ipfs/go-ipfs/releases) and upload the result of the ipfs/distributions build in the previous step.
- Announce the RC:
  - [ ] On IRC/Matrix (both #ipfs and #ipfs-dev)
  - [ ] To the _early testers_ listed in [docs/EARLY_TESTERS.md](https://github.com/ipfs/go-ipfs/tree/master/docs/EARLY_TESTERS.md).

Checklist:

- [x] **Stage 0 - Automated Testing**
  - [x] Fork a new branch (`release-vX.Y.Z`) from `master` and make any further release related changes to this branch. If any "non-trivial" changes (see the footnotes of [docs/releases.md](https://github.com/ipfs/go-ipfs/tree/master/docs/releases.md) for a definition) get added to the release, uncheck all the checkboxes and return to this stage.
    - [x] Follow the RC release process to cut the first RC.
    - [x] Bump the version in `version.go` in the `master` branch to `vX.(Y+1).0-dev`.
  - [x] Automated Testing (already tested in CI) - Ensure that all tests are passing, this includes:
    - [x] unit, sharness, cross-build, etc (`make test`)
    - [x] lint (`make test_go_lint`)
    - [x] [interop](https://github.com/ipfs/interop#test-with-a-non-yet-released-version-of-go-ipfs)
    - [x] [go-ipfs-api](https://github.com/ipfs/go-ipfs-api)
    - [x] [go-ipfs-http-client](https://github.com/ipfs/go-ipfs-http-client)
    - [x] [WebUI](https://github.com/ipfs-shipyard/ipfs-webui)
- [x] **Stage 1 - Internal Testing**
  - [x] CHANGELOG.md has been updated
    - use [`./bin/mkreleaselog`](https://github.com/ipfs/go-ipfs/tree/master/bin/mkreleaselog) to generate a nice starter list
  - [x] Infrastructure Testing:
    - [x] Deploy new version to a subset of Bootstrappers
    - [x] Deploy new version to a subset of Gateways
    - [x] Deploy new version to a subset of Preload nodes
    - [x] Collect metrics every day. Work with the Infrastructure team to learn of any hiccup
  - [x] IPFS Application Testing -  Run the tests of the following applications:
    - [x] [IPFS Desktop](https://github.com/ipfs-shipyard/ipfs-desktop) - @lidel
    - [x] [IPFS Companion](https://github.com/ipfs-shipyard/ipfs-companion) - @lidel
- [x] **Stage 2 - Community Dev Testing**
  - [x] Reach out to the IPFS _early testers_ listed in [docs/EARLY_TESTERS.md](https://github.com/ipfs/go-ipfs/tree/master/docs/EARLY_TESTERS.md) for testing this release (check when no more problems have been reported). If you'd like to be added to this list, please file a PR.
  - [x] Reach out to on IRC for beta testers.
  - [x] Run tests available in the following repos with the latest beta (check when all tests pass):
    - [x] [orbit-db](https://github.com/orbitdb/orbit-db)
- [x] **Stage 3 - Community Prod Testing**
  - [x] Documentation
    - [x] Ensure that [CHANGELOG.md](https://github.com/ipfs/go-ipfs/tree/master/CHANGELOG.md) is up to date
    - [x] Ensure that [README.md](https://github.com/ipfs/go-ipfs/tree/master/README.md)  is up to date
    - [x] Ensure that all the examples we have produced for go-ipfs run without problems
    - [x] Update HTTP-API Documentation on the Website using https://github.com/ipfs/http-api-docs
    - [x] Update [CLI Documentation](https://github.com/ipfs/ipfs-docs/blob/master/docs/reference/cli.md) on the Website using https://github.com/ipfs-inactive/docs/blob/master/scripts/cli.sh
  - [x] Invite the wider community through (link to the release issue):
    - [x] Matrix
    - [x] Discuss
- [x] **Stage 4 - Release**
  - [x] Final preparation
    - [x] Verify that version string in [`version.go`](https://github.com/ipfs/go-ipfs/tree/master/version.go) has been updated.
    - [x] Merge `release-vX.Y.Z` into the `release` branch.
    - [x] Tag this merge commit (on the `release` branch) with `vX.Y.Z`.
    - [x] Release published
      - [x] to [dist.ipfs.io](https://dist.ipfs.io)
      - [x] to [npm-go-ipfs](https://github.com/ipfs/npm-go-ipfs)
      - [x] to [chocolatey](https://chocolatey.org/packages/ipfs)
      - [x] to [snap](https://snapcraft.io/ipfs)
      - [x] to [github](https://github.com/ipfs/go-ipfs/releases)
      - [x] to [arch](https://www.archlinux.org/packages/community/x86_64/go-ipfs/) (flag it out of date)
    - [x] Cut a new ipfs-desktop release
  - [x] Publish a Release Blog post (at minimum, a c&p of this release issue with all the highlights, API changes, link to changelog and thank yous)
  - [x] Broadcasting (link to blog post)
    - [x] Twitter
    - [x] Matrix
    - [x] [Reddit](https://reddit.com/r/ipfs)
    - [x] [discuss.ipfs.io](https://discuss.ipfs.io/c/announcements)
    - [x] ~Announce it on the [IPFS Users Mailing List](https://groups.google.com/forum/#!forum/ipfs-users)~
- [x] **Post-Release**
  - [x] Merge the `release` branch back into `master`, ignoring the changes to `version.go` (keep the `-dev` version from master).
  - [x] Create an issue using this release issue template for the _next_ release.
  - [x] Make sure any last-minute changelog updates from the blog post make it back into the CHANGELOG.

# ⁉️ Do you have questions?

The best place to ask your questions about IPFS, how it works and what you can do with it is at [discuss.ipfs.io](http://discuss.ipfs.io). We are also available at the [`#lobby:ipfs.io`](https://matrix.to/#/#lobby:ipfs.io) Matrix channel which is bridged with [other chat platforms](https://docs.ipfs.io/community/chat/).

**Comment by eminence** [comment]
> improvements that will enable us to release faster with more effort.

more effort, or less effort?

**Comment by BigLep** [comment]
@eminence : doh - fixed - thanks!

**Comment by aschmahmann** [comment]
## Changelog

< changelog generated by bin/mkreleaselog >

## ❤️ Contributors

< list generated by bin/mkreleaselog >

Would you like to contribute to the IPFS project and don't know how? Well, there are a few places you can get started:

- Check the issues with the `help wanted` label in the [go-ipfs repo](https://github.com/ipfs/go-ipfs/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22)
- Join an IPFS All Hands, introduce yourself and let us know where you would like to contribute - https://github.com/ipfs/team-mgmt/#weekly-ipfs-all-hands
- Hack with IPFS and show us what you made! The All Hands call is also the perfect venue for demos, join in and show us what you built
- Join the discussion at [discuss.ipfs.io](https://discuss.ipfs.io/) and help users finding their answers.
- Join the [🚀 IPFS Core Implementations Weekly Sync 🛰](https://github.com/ipfs/team-mgmt/issues/992) and be part of the action!

**Comment by ianopolous** [comment]
My reading of this is that this doesn't include the pending move of the blockstore from cids to multihashes. Is that correct?

**Comment by aschmahmann** [comment]
@ianopolous correct. That is currently planned to be a release all of its own (i.e. that will be the only change in that release in order to ease people through the migration). It's currently scheduled for v0.12.0

**Comment by b5** [comment]
Is anyone aware of anything like https://deno.land/benchmarks for go-ipfs? The switch to go-ipld-prime under the hood is very exciting (thanks @hannahhoward !). To help assess release candidates, I'd really appreciate the sanity checks an automated set of benchmarks run between releases would provide.

**Comment by aschmahmann** [comment]
Early testers ping for testing 😄

- [ ] pacman.store ([@RubenKelevra](https://github.com/RubenKelevra))
- [ ] Infura (@MichaelMure)
- [ ] Textile (@sanderpick)
- [ ] Pinata (@obo20)
- [ ] RTrade (@postables)
- [ ] QRI (@b5)
- [ ] Siderus (@koalalorenzo)
- [ ] Charity Engine (@rytiss, @tristanolive)
- [ ] Fission (@bmann)

**Comment by Jorropo** [comment]
> <List of items with PRs and/or Issues to be considered for this release>

#8177 pls ?

**Comment by obo20** [comment]
> Early testers ping for RC testing 😄
> 
> * [ ]  pacman.store ([@RubenKelevra](https://github.com/RubenKelevra))
> * [ ]  Infura (@MichaelMure)
> * [ ]  Textile (@sanderpick)
> * [x]  Pinata (@obo20)
> * [ ]  RTrade (@postables)
> * [ ]  QRI (@b5)
> * [ ]  Siderus (@koalalorenzo)
> * [ ]  Charity Engine (@Rytiss, @tristanolive)
> * [ ]  Fission (@bmann)

Things are looking good on our end. We've been running this on a segment of our infrastructure and there are no issues to report as of yet.

**Comment by tabcat** [comment]
tracking orbit-db support: https://github.com/orbitdb/orbit-db/issues/909

**Comment by kallisti5** [comment]
Any thoughts on #8413 ?  Seems like an easy add with a big usability improvement.

**Comment by BigLep** [comment]
Per 2021-09-09 verbal:
1. Clearer callout on the breaking changes TLDR (e.g., ipfs-dag break)
2. Clearer documentation on how to migrate 
3. Need to call out determinstic CBOR encoding
4. Need to add callouts for the various items Rod found
5. We need to track down that the JS HTTP tests didn't run (related to OrbitDB also finding failures during their test runs).  This may just need to be a tracking issue for cases where Go is ahead of JS.

**Comment by lidel** [comment]
## v0.10.0-rc2

Just published RC2 on [github](https://github.com/ipfs/go-ipfs/releases/tag/v0.10.0-rc2), [dist.ipfs.io](https://dist.ipfs.io/go-ipfs/v0.10.0-rc2), NPM and [dockerhub](https://hub.docker.com/r/ipfs/go-ipfs/tags/?page=1&ordering=last_updated&name=v0.10.0-rc2).

:point_right: We are planning to ship the final release around Thursday 30th (TBD), mind the window for testing is a bit short.

## Changes since -rc1

- improved how we introduce breaking changes  in `ipfs dag get` and `ipfs dag put`
  - rc1 had them too, but we realized that some edge cases around codec name aliases were difficult to reason about and could lead to silent errors
  - rc2 makes the changes very clear and explicit (renamed params, strict codec names). this  should make migration and refactoring of userland apps that rely on the old behavior of  `dag pub/get` easier with no surface for ambiguity
  - details in [the updated draft of release notes](https://github.com/ipfs/go-ipfs/issues/8176#issue-911985107)
- new commands / parameters:
  - `ipfs multibase transcode --help`
  - `ipfs swarm peering --help`
  - `ipfs dag import --stats`

## Call for feedback

[Early testers](https://github.com/ipfs/go-ipfs/tree/master/docs/EARLY_TESTERS.md) – ping for `v0.10.0-rc2`  testing 😄

- [ ] pacman.store ([@RubenKelevra](https://github.com/RubenKelevra))
- [ ] Infura (@MichaelMure)
- [ ] Textile (@sanderpick)
- [ ] Pinata (@obo20)
- [ ] RTrade (@postables)
- [ ] QRI (@b5)
- [ ] Siderus (@koalalorenzo)
- [ ] Charity Engine (@rytiss, @tristanolive)
- [ ] Fission (@bmann)
- [ ] OrbitDB (@tabcat)
- [ ] Companion/Desktop (@lidel)

**Comment by MichaelMure** [comment]
Would it be reasonable to include [this merkledag fix](https://github.com/ipfs/go-merkledag/pull/75) ?

**Comment by tabcat** [comment]
tests are passing locally using https://github.com/orbitdb/orbit-db/issues/913 (will be merged and published very soon) + go-ipfs@0.10.0-rc2. :+1:

**Comment by aschmahmann** [comment]
@MichaelMure IIUC that fix doesn't effect really effect the go-ipfs binary, but might be useful in some edge cases with go-ipfs as a library. 

If so then we can do a go-merkledag release. I'd prefer not to add much more into the release if we can to expedite getting it out.

If there's a more pressing need or problem that I'm missing lmk (here, on the go-merkeldag PR, or in a DM)
